packages/clay-modal/src/Provider.tsx

```diff
@@ -0,0 +1,91 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayModal from './Modal';
+import React, {createContext, useReducer} from 'react';
+import {Size, Status} from './types';
+
+enum Action {
+	Close = 0,
+	Open = 1,
+}
+
+interface IProps {
+	spritemap?: string;
+}
+
+type TState = {
+	body: React.ReactElement | React.ReactText;
+	footerFirst?: React.ReactElement;
+	footerLast?: React.ReactElement;
+	header?: React.ReactElement | React.ReactText;
+	size?: Size;
+	status?: Status;
+	url?: string;
+};
+
+type TAction = {type: Action.Open; payload: TState} | {type: Action.Close};
+
+type TProvider = [TState, React.Dispatch<TAction>];
+
+const initialState = {
+	body: <div />,
+	visible: false,
+};
+
+const reducer = (
+	state: TState,
+	action: TAction
+): TState & {visible: boolean} => {
+	switch (action.type) {
+		case Action.Open:
+			return {...action.payload, visible: true};
+		case Action.Close:
+			return initialState;
+		default:
+			throw new TypeError();
+	}
+};
+
+const Context = createContext<Partial<TProvider>>([]);
+
+const ClayModalProvider: React.FunctionComponent<IProps> = ({
+	children,
+	spritemap,
+}) => {
+	const [{visible, ...state}, dispatch] = useReducer(reducer, initialState);
+	const {body, footerFirst, footerLast, header, size, status, url} = state;
+
+	return (
+		<>
+			{visible && (
+				<ClayModal
+					onClose={() => dispatch({type: Action.Close})}
+					size={size}
+					spritemap={spritemap}
+					status={status}
+				>
+					{onClose => (
```

Comment:
Woops, just seeing this now. I noted below about this idea as well. I think its a great idea to use onClose via hook or context instead of renderProps

Reply:
Oh great, We are aligned. 🙂